File: scripts/events_to_csv.rb
require "nokogiri"
require 'open-uri'
require 'date'
require_relative 'functions.rb'

all_the_matches = File.readlines("data/matches.csv").
                       map { |row| row.gsub(/\n/, "").split(",") }

MATCHES = {}

all_the_matches.each do |match|
	match_id = match[5].gsub(/\.html/, "").split("/")[-2].split("=")[1]
	MATCHES[match_id] = match[0]
end

COUNTRY_CODES = {}
cc = File.readlines("data/cc.csv").
                       map { |row| row.gsub(/\n/, "").split(",") }

cc.each {|k, v| COUNTRY_CODES[k] = v.strip }

all_the_matches = File.readlines("data/matches.csv").
                       map { |row| row.gsub(/\n/, "").split(",") }

EVENTS = []

player_id_lookup = Functions.player_id_lookup

match_files = Dir["data/matches/*"]
match_files.select {|file| !file.end_with?(".md") && !file.include?("stats") }.each do |match_file|
	match_id = match_file.gsub(/\.html/, "").split("/")[-1]
	year = MATCHES[match_id]
	puts "processing #{match_id}"
	
	@doc = Nokogiri::HTML(open(match_file))	
	world_cup = @doc.css("div.content_header h1").text

	if year == "2010"
		@doc = Nokogiri::HTML(open("data/matches/stats/#{match_id}.html"))	

		@doc.css("div.scorers ul.shome li").each do |scorer|
			if Functions.multiple_goals?(scorer.text)
				goals = Functions.process_multiple_scorer(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])
				goals.each do |goal|
					EVENTS << {:match_id => match_id}.merge(goal)
				end
			else
				vals = Functions.process_scorer(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])
				EVENTS << {:match_id => match_id}.merge(vals)
			end			
		end

		@doc.css("div.scorers ul.saway li").each do |scorer|
			if Functions.multiple_goals?(scorer.text)
				goals = Functions.process_multiple_scorer(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])
				goals.each do |goal|
					EVENTS << {:match_id => match_id}.merge(goal)
				end
			else
				vals = Functions.process_scorer(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])
				EVENTS << {:match_id => match_id}.merge(vals)
			end	
		end		

		@doc.css("table.cards ul li").each do |row|
			player = row.css("span:nth-child(2)")

			if player.count == 0
				player = row.css("div")
			end

			yellow = row.css("span.icon-events-yellow")
			yellowred = row.css("span.icon-events-yellow2red")
			red = row.css("span.icon-events-red")

			vals = Functions.process_card(player.text, COUNTRY_CODES, player_id_lookup[year][:all])

			if yellow.count > 0 
				EVENTS << {:match_id => match_id, :type => "yellow"}.merge(vals)
			end

			if yellowred.count > 0 
				EVENTS << {:match_id => match_id, :type => "yellowred"}.merge(vals)
			end

			if red.count > 0 
				EVENTS << {:match_id => match_id, :type => "red"}.merge(vals)
			end			
			
		end		
	else
		scorers = @doc.css("div.cont:nth-child(4) li")

		current_scorer = nil
		scorers.each do |scorer|
			current_scorer = scorer.text if current_scorer.nil?
			if scorer.text.start_with? " "
				time = scorer.text.match(/(\d{1,3})'/)[1]
				current_scorer = current_scorer.gsub(/\d{1,3}/, time)
			else
				current_scorer = scorer.text
			end			

			vals = Functions.process_scorer(current_scorer, COUNTRY_CODES, player_id_lookup[year][:all])
			EVENTS << {:match_id => match_id}.merge(vals)
		end	

		yellows = @doc.css("div.cont:nth-child(6) li")

		yellows.each do |scorer|
			next if scorer.text.start_with? ", "
			vals = Functions.process_card(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])

			EVENTS << {:match_id => match_id, :type => "yellow"}.merge(vals)
		end

		reds = @doc.css("div.cont:nth-child(7) li")

		reds.each do |scorer|
			vals = Functions.process_card(scorer.text, COUNTRY_CODES, player_id_lookup[year][:all])
			EVENTS << {:match_id => match_id, :type => "red"}.merge(vals)
		end	
	end
end

open("data/import/events.csv", 'w') { |f|	
	f.puts (["match_id", 
			 "player",
			 "player_id",
			 "time",
			 "type"]).join(",")

  	EVENTS.each do |event|
		f.puts ["\"#{event[:match_id]}\"",
		    	"\"#{event[:player]}\"",
		    	"\"#{event[:player_id]}\"",
				"#{event[:time]}",
				"\"#{event[:type]}\""
				].join(",") 
	end
}

File: scripts/python/events_to_csv.py
import csv
import glob
import json

from bs4 import BeautifulSoup
import re

with open("data/2018/import/events.csv", "w") as matches_file:
    writer = csv.writer(matches_file, delimiter=",")

    writer.writerow([
        "match_id", "player_id", "time", "type", "scoring_team_id"
    ])

    for match_file in glob.glob("data/2018/matches/*.html"):
        print(match_file)
        match_id = match_file.split("/")[-1].replace(".html", "")
        with open(match_file, "r") as file:
            soup = BeautifulSoup(file.read(), "html.parser")

            header_element = soup.select("div h1 a")

            if len(header_element) == 0:
                header_element = soup.select("div a h1")

            header = header_element[0].text
            groups = re.match("(\d{4}) FIFA World Cup ([a-zA-Z ]*)", header)
            year = int(groups[1])

            if year == 2018:
                with open("data/2018/matches/api/{0}.json".format(match_id)) as json_file:
                    the_json = json.loads(json_file.read())

                    for scorer in the_json["HomeTeam"]["Goals"]:
                        type = scorer["Type"]
                        event_time = scorer["Minute"]
                        event_type = "goal"
                        if type == 1:
                            event_type = "penalty"
                            if scorer["Period"] == 11:
                                event_type = "penalty-shootout"
                        if type == 3:
                            event_type = "owngoal"

                        player_id = scorer["IdPlayer"]

                        scoring_team_id = the_json["HomeTeam"]["IdTeam"]

                        values = [
                            match_id, player_id, event_time, event_type, scoring_team_id
                        ]

                        values = [str(value).strip() for value in values]
                        writer.writerow(values)

                    for scorer in the_json["AwayTeam"]["Goals"]:
                        type = scorer["Type"]
                        event_time = scorer["Minute"]
                        event_type = "goal"
                        if type == 1:
                            event_type = "penalty"
                            if scorer["Period"] == 11:
                                event_type = "penalty-shootout"
                        if type == 3:
                            event_type = "owngoal"

                        player_id = scorer["IdPlayer"]

                        scoring_team_id = the_json["AwayTeam"]["IdTeam"]

                        values = [
                            match_id, player_id, event_time, event_type, scoring_team_id
                        ]

                        values = [str(value).strip() for value in values]
                        writer.writerow(values)
            else:
                scorers_teams = soup.select("ul.mh-l-scorers")

                for scoring_team in scorers_teams:
                    scoring_team_id = scoring_team["data-team-id"]
                    scorers = scoring_team.select("li.mh-scorer")
                    for scorer in scorers:
                        scorer_element = scorer.select("div.p-i-no")[0]
                        player_id = scorer_element["data-player-id"]
                        events = scorer.select("span.ml-scorer-evmin span")
                        for event in events:
                            event_time = event.text.replace(",", "")

                            event_type = "goal"
                            if "PEN" in event_time:
                                event_type = "penalty"
                            if "OG" in event_time:
                                event_type = "owngoal"

                            event_time = event_time.replace("OG", "").replace("PEN", "").strip()

                            values = [
                                match_id, player_id, event_time, event_type, scoring_team_id
                            ]
                            values = [str(value).strip() for value in values]
                            writer.writerow(values)

                        # events = soup.select("div.p-lu-info")
                        # for event in events:
                        #     yellow_cards = event.select("span.p-e span.yellow-card")
